zanata-client-commands/src/main/java/org/zanata/client/commands/pull/PullCommand.java

```diff
@@ -330,6 +287,139 @@ public void run() throws Exception {
+            if (transFile.exists()
+                    && Long.toString(transFile.lastModified())
+                            .equals(eTagCacheEntry
+                                    .getLocalFileTime())) {
+                eTag = eTagCacheEntry.getServerETag();
+            }
+        }
+
+        ClientResponse transResponse =
+                transDocResourceClient.getTranslations(docUri,
+                        locale, strat.getExtensions(),
+                        createSkeletons, eTag);
+
+        // ignore 404 (no translation yet for specified
+        // document)
+        if (transResponse.getClientResponseStatus() == ClientResponse.Status.NOT_FOUND) {
+            if (!createSkeletons) {
+                log.info(
+                        "No translations found in locale {} for document {}",
+                        locale, localDocName);
+            } else {
+                // Write the skeleton
+                writeTargetDoc(strat, localDocName, locMapping,
+                    doc, null,
+                    transResponse.getHeaders()
+                        .getFirst(HttpHeaders.ETAG));
+            }
+        } else if (transResponse.getClientResponseStatus() == ClientResponse.Status.NOT_MODIFIED) {
+            // 304 NOT MODIFIED (the document can stay the same)
+            log.info(
+                    "No changes in translations for locale {} and document {}",
+                    locale, localDocName);
+
+            // Check the file's MD5 matches what's stored in the
+            // cache. If not, it needs to be fetched again (with
+            // no etag)
+            String fileChecksum =
+                    HashUtil.getMD5Checksum(transFile);
+            if (!fileChecksum.equals(eTagCacheEntry
+                    .getLocalFileMD5())) {
+                transResponse =
+                        transDocResourceClient.getTranslations(
+                                docUri, locale,
+                                strat.getExtensions(),
+                                createSkeletons, null);
+                ClientUtil.checkResult(transResponse);
+                // rewrite the target document
+                writeTargetDoc(strat, localDocName, locMapping,
+                    doc, transResponse.getEntity(TranslationsResource.class),
+                    transResponse.getHeaders()
+                        .getFirst(HttpHeaders.ETAG));
+            }
+        } else {
+            ClientUtil.checkResult(transResponse);
+            TranslationsResource targetDoc =
+                transResponse.getEntity(TranslationsResource.class);
+
+            // Write the target document
+            writeTargetDoc(strat, localDocName, locMapping,
+                    doc, targetDoc,
+                    transResponse.getHeaders()
+                            .getFirst(HttpHeaders.ETAG));
+        }
+    }
+
+    private boolean shouldPullThisLocale(
+            Map<String, Map<LocaleId, TranslatedPercent>> statsMap,
```

Comment:
`@Nullable` would be even better to have here (unless you want to use Optional).